contracts/src/Lightning.ts

```diff
@@ -51,19 +52,26 @@ export const RecursiveProgram = Experimental.ZkProgram({
     },
 
     step: {
-      privateInputs: [SelfProof],
+      privateInputs: [SelfProof, Signature],
 
       /**
        * this method only cares about publicInput.transferFrom1To2
        * notice that transferFrom1To2 can be negative depending on who is transferring to who
        */
-      method(publicInput: RecursivePublicInput, earlierProof: SelfProof<RecursivePublicInput>) {
+      method(publicInput: RecursivePublicInput, earlierProof: SelfProof<RecursivePublicInput>, senderSignature: Signature) {
         // verify earlier proof
         earlierProof.verify();
         // assert balances are >= 0 for both parties
         publicInput.user1Balance.assertGreaterThanOrEqual(0, "user1 balance cannot be < 0 due to this transfer")
         publicInput.user2Balance.assertGreaterThanOrEqual(0, "user2 balance cannot be < 0 due to this transfer")
 
+        // require the public input to be signed by the asset sender
+        Circuit.if(
+          publicInput.transferFrom1to2.greaterThan(0),
+          senderSignature.verify(publicInput.user1, RecursivePublicInput.toFields(publicInput)),
+          senderSignature.verify(publicInput.user2, RecursivePublicInput.toFields(publicInput))
+        )
```

Comment:
Awesome addition!

Reply:
:)

Reply:
Actually, #3 was to point out this fact; anyone who knows the current proof state in the channel can move the asset freely. I.e., the channel participant can move all asset of other to theirs; and this is the fix for that.